rfc/20250721-new-plugin-protocol.md

```diff
@@ -0,0 +1,361 @@
+  {                           // Parameters
+    "resources": [...],
+    "config": {...}
+  }
+]
+
+// Response - actually sent as: [1, 1, null, {"warnings": [...]}]
+[
+  1,                          // Message type 1 = Response
+  1,                          // Message ID (matches request)
+  null,                       // Error (null = no error)
+  {                           // Result
+    "warnings": ["Resource cost exceeds $1000/month"]
+  }
+]
+
+// Notification - actually sent as: [2, "log", {"level": "info", ...}]
+[
+  2,                          // Message type 2 = Notification
+  "log",                      // Method name
+  {                           // Parameters
+    "level": "info",
+    "message": "Processing resource aws_instance.example"
+  }
+]
+```
+
+#### Why stdio?
+
+Standard input/output is the universal IPC mechanism:
+- Every language can read stdin and write stdout
+- No network ports or sockets to manage
+- Works everywhere - containers, restricted environments, etc.
+- Simple to debug and test
+
+#### Message Transport
+
+Following Neovim's proven approach, we don't need explicit framing. MessagePack is self-delimiting - each value contains enough information to determine where it ends. This makes the protocol simpler:
+
+1. **Writing**: Encode msgpack-rpc arrays directly to stdout
+2. **Reading**: Use a streaming msgpack parser that:
+    - Reads bytes from stdin into a buffer
+    - Feeds them to the msgpack unpacker
+    - Emits complete messages as they're decoded
+    - Handles partial messages automatically
+
+This approach has several advantages:
+- **Simplicity**: No framing headers to implement or parse
+- **Proven**: Neovim's large plugin ecosystem demonstrates this works well
+- **Library support**: Most msgpack libraries include streaming parsers
+- **Resilient**: Parser can recover from corrupted data
+
+Example pseudo-code for reading:
+```python
+unpacker = msgpack.Unpacker()
+while True:
+    data = stdin.read(4096)  # Read available bytes
+    unpacker.feed(data)
+    for message in unpacker:  # Yields complete messages
+        process_msgpack_rpc_message(message)
+```
+
+### Message Types (Proposed)
+
+Beyond the standard msgpack-rpc message types, we propose the following methods for plugin lifecycle:
+
+1. **Initialization**
+    - `init`: Plugin initialization and capability declaration
+    - Returns: Plugin metadata and supported capabilities
+
+2. **Plugin-specific methods**
+    - Methods vary based on declared capabilities
+    - Examples: `middleware/prePlan`, `storage/read`, `functions/call`
+
+3. **Common methods**
+    - `shutdown`: Graceful termination
+    - `ping`: Health check
+
+### Plugin Types and Capabilities
+
+We propose a unified plugin SDK that supports multiple capability levels. A single plugin can declare multiple capabilities, allowing natural evolution from simple to complex functionality:
+
+#### Level 1: Middleware Capabilities
```

Comment:
I think we could get caught up in the specifics of each of these capabilities in this RFC.

If I understand correctly, these are more so examples of what it could look like and where we might start?

Reply:
I don't feel _super_ strongly about this, but I think it could be helpful to deal with provider-defined functions as an earlier step to give us some experience with adapting features from the existing protocol, in case we learn something doing that which causes us to want to tweak some of the design details.

Provider-defined functions are relatively simple compared to most other things in the existing provider protocol, so I'd personally suggest we try to tackle them _first_. We could continue to work on the overall design for "middleware" concurrently with this work and then tackle it as the second feature, rather than trying to introduce a new protocol and an entirely new feature at the same time.

:thinking: